src/test/scala/akka/stream/contrib/RetrySpec.scala

```diff
@@ -0,0 +1,264 @@
+
+      sink.request(99)
+      source.sendNext(1)
+      assert(sink.expectNext()._1 === Success(2))
+      source.sendNext(2)
+      assert(sink.expectNext()._1 === Success(4))
+      assert(sink.expectNext()._1 === Success(2))
+      assert(sink.expectNext()._1 === Success(2))
+      source.sendNext(44)
+      assert(sink.expectNext()._1 === failedElem)
+      source.sendNext(42)
+      sink.expectNoMsg()
+      source.sendComplete()
+      sink.expectComplete()
+    }
+
+    "retry squares by division" in {
+      val (source, sink) = TestSource.probe[Int]
+        .map(i => (i, i * i))
+        .via(Retry.concat(100, flow[Int]) {
+          case x if x % 4 == 0 => Some(List((x / 2, x / 4)))
+          case x => {
+            val sqrt = scala.math.sqrt(x.toDouble).toInt
+            Some(List((sqrt, x)))
+          }
+        })
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+      source.sendNext(1)
+      assert(sink.expectNext()._1 === Success(2))
+      source.sendNext(2)
+      assert(sink.expectNext()._1 === Success(2))
+      source.sendNext(4)
+      assert(sink.expectNext()._1 === Success(2))
+      sink.expectNoMsg(3.seconds)
+      source.sendComplete()
+      sink.expectComplete()
+    }
+
+    "tolerate killswitch terminations after start" in {
+      val ((source, killSwitch), sink) = TestSource.probe[Int]
+        .viaMat(KillSwitches.single[Int])(Keep.both)
+        .map(i => (i, i * i))
+        .via(Retry.concat(100, flow[Int]) {
+          case x if x % 4 == 0 => Some(List((x / 2, x / 4)))
+          case x => {
+            val sqrt = scala.math.sqrt(x.toDouble).toInt
+            Some(List((sqrt, x)))
+          }
+        })
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+      source.sendNext(1)
+      assert(sink.expectNext()._1 === Success(2))
+      source.sendNext(2)
+      assert(sink.expectNext()._1 === Success(2))
+      killSwitch.abort(failedElem.failed.get)
+      sink.expectError(failedElem.failed.get)
+    }
+
+    "tolerate killswitch terminations on start" in {
+      val ((source, killSwitch), sink) = TestSource.probe[Int]
+        .viaMat(KillSwitches.single[Int])(Keep.both)
+        .map(i => (i, i))
+        .via(Retry.concat(100, flow[Int]) { x => Some(List((x, x + 1))) })
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      sink.request(99)
+      killSwitch.abort(failedElem.failed.get)
+      sink.expectError(failedElem.failed.get)
+    }
+
+    "tolerate killswitch terminations before start" in {
+      val ((source, killSwitch), sink) = TestSource.probe[Int]
+        .viaMat(KillSwitches.single[Int])(Keep.both)
+        .map(i => (i, i))
+        .via(Retry.concat(100, flow[Int]) { x => Some(List((x, x + 1))) })
+        .toMat(TestSink.probe)(Keep.both)
+        .run()
+
+      killSwitch.abort(failedElem.failed.get)
+      sink.request(1)
+      sink.expectError(failedElem.failed.get)
+    }
```

Comment:
I think it would be important to have a version where the flow passed to Retry itself is terminated via a KillSwitch